### PR TITLE
Creating a user with an empty password will send an email to let him set his password

### PR DESCRIPTION
**Pull request by JohnXLivingston**
With this PR, an administrator will be able to create users without entering their password. 
Prerequisite: emailing must be enabled (smtp configured).
When an administrator leave the password field blank, then the code will generate a random password (20 character long), and send an email to the user. This email contains a reset password link, valid for 7 days.

NB: the create user api apparently can be called by non-privileged users. I didn't block the ability for non-administrators to use this feature. 

NB: I haven't updated the documentation or added unit tests. Let me know if this PR is ok for you, and I will do this.

### CODE REVIEW

File: client/src/app/+admin/users/user-edit/user-create.component.ts [hunk 1]
```diff
@@ -42,10 +43,13 @@ export class UserCreateComponent extends UserEdit implements OnInit {
       videoQuotaDaily: '-1'
     }
 
+    const serverConfig = this.route.snapshot.data.serverConfig
+    const isEmailEnabled = !!serverConfig.email.enabled
```
Comment (Chocobozzz) [inline client/src/app/+admin/users/user-edit/user-create.component.ts]: It's already a boolean

File: client/src/app/shared/forms/form-validators/user-validators.service.ts [hunk 2]
```diff
@@ -56,6 +57,17 @@ export class UserValidatorsService {
       }
     }
 
+    this.USER_PASSWORD_OPTIONAL = {
```
Comment (Chocobozzz) [inline client/src/app/shared/forms/form-validators/user-validators.service.ts]: Why don't you use `USER_PASSWORD`?

File: server/helpers/custom-validators/users.ts [hunk 3]
```diff
@@ -3,13 +3,22 @@ import { UserRole } from '../../../shared'
...
 const USERS_CONSTRAINTS_FIELDS = CONSTRAINTS_FIELDS.USERS
 
 function isUserPasswordValid (value: string) {
   return validator.isLength(value, USERS_CONSTRAINTS_FIELDS.PASSWORD)
 }
 
+function isUserPasswordValidOrEmpty (value: string) {
+  // Empty password is only possible if emailing is enabled.
+  if (value === '') {
+    return !!Emailer.isEnabled()
```
Comment (Chocobozzz) [inline server/helpers/custom-validators/users.ts]: This is already a boolean

File: server/controllers/api/users/index.ts [hunk 4]
```diff
@@ -197,11 +197,25 @@ async function createUser (req: express.Request, res: express.Response) {
     adminFlags: body.adminFlags || UserAdminFlag.NONE
   }) as MUser
 
+  // NB: due to the validator usersAddValidator, password==='' can only be true if we can send the mail.
+  const createPassword = userToCreate.password === ''
```
Comment (Chocobozzz) [inline server/controllers/api/users/index.ts]: `if (!userToCreate.password) userToCreate.password = await generateRandomString(20)` should be enough, and I don't think the comment is necessary

File: server/lib/emailer.ts [hunk 5]
```diff
@@ -384,6 +384,22 @@ class Emailer {
     return JobQueue.Instance.createJob({ type: 'email', payload: emailPayload })
   }
 
+  addPasswordCreateEmailJob (username: string, to: string, resetPasswordUrl: string) {
+    const text = 'Hi,\n\n' +
+      `Welcome to your ${WEBSERVER.HOST}'s account. Your username is: ${username}.\n\n` +
```
Comment (Chocobozzz) [inline server/lib/emailer.ts]: `Welcome to your ${WEBSERVER.HOST} PeerTube instance.`

File: server/lib/emailer.ts [hunk 6]
```diff
@@ -384,6 +384,22 @@ class Emailer {
...
+      `Please set your password by following this link: ${resetPasswordUrl} (this link will expire within seven days).\n\n` +
+      'Cheers,\n' +
+      `${CONFIG.EMAIL.BODY.SIGNATURE}`
+
+    const emailPayload: EmailPayload = {
+      to: [ to ],
+      subject: CONFIG.EMAIL.SUBJECT.PREFIX + 'Set your password',
```
Comment (Chocobozzz) [inline server/lib/emailer.ts]: `New PeerTube account password`

**Comment by JohnXLivingston** [comment]
Thanks for the review. I will check that as soon as I have some time! (maybe tonight or tomorrow)

File: client/src/app/shared/forms/form-validators/user-validators.service.ts (hunk 2 above)
Comment (JohnXLivingston) [reply client/src/app/shared/forms/form-validators/user-validators.service.ts]: I have to remove the require constraint.

File: server/controllers/api/users/index.ts (hunk 4 above)
Comment (JohnXLivingston) [reply server/controllers/api/users/index.ts]: I need the createPassword vars on line 211, to send the email

**Comment by JohnXLivingston** [comment]
> Please could you add a small text below the password field saying that if it is not filled by the admin, an email will be sent to the user?

For localization, have I something more than adding an i18n div to do? Have I to declare the sentence somewhere?

**Comment by JohnXLivingston** [comment]
I wanted to do the unit tests, but I am having some difficulties. I'm not at all familiar with mocha, and I keep getting the following error:


`npm run mocha -- --exit -r ts-node/register -r tsconfig-paths/register --bail server/tests/api/check-params/users.ts 
[...]
  Test users API validators
    1) "before all" hook in "Test users API validators"


  0 passing (7s)
  1 failing

  1) Test users API validators
       "before all" hook in "Test users API validators":
     Error: expected 200 "OK", got 400 "Bad Request"
      at Test._assertStatus (node_modules/supertest/lib/test.js:268:12)
      at Test._assertFunction (node_modules/supertest/lib/test.js:283:11)
      at Test.assert (node_modules/supertest/lib/test.js:173:18)
      at localAssert (node_modules/supertest/lib/test.js:131:12)
      at /home/john/dev/PeerTube/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1145:12)
      at process._tickCallback (internal/process/next_tick.js:63:19)

`


Same as here for example: https://travis-ci.org/Chocobozzz/PeerTube/jobs/650876900?utm_medium=notification&utm_source=github_status

I have no clue on what it is. I don't think it is related to my PR, but I don't know what to do. I don't understand which request fails, so I can't try to fix the issue.

**Comment by JohnXLivingston** [comment]
> I have no clue on what it is. I don't think it is related to my PR, but I don't know what to do. I don't understand which request fails, so I can't try to fix the issue.

Finally it seems that my code is responsible. But I don't understand why. I'm investigating.

**Comment by JohnXLivingston** [comment]
> > I have no clue on what it is. I don't think it is related to my PR, but I don't know what to do. I don't understand which request fails, so I can't try to fix the issue.
> 
> Finally it seems that my code is responsible. But I don't understand why. I'm investigating.

Ok... After 2h of work I find how to debug server side code (I'm totally new to node), and I found the mistake... It seems to be a circular imports problems... Importing server/lib/emailer in server/helpers/custom-validators/users does not work. And it don't generate any error on compile time.
It is fixed in commit https://github.com/JohnXLivingston/PeerTube/commit/2c49689edb10f53f69fdd8d26e2d5c78148c5ed2 . But I am not very happy with this fix... If Emailer.isEnabled changes its code, it can make a divergence.

**Comment by JohnXLivingston** [comment]
I just comited the unit tests. For my this PR is finished.

**Comment by Chocobozzz** [comment]
Thanks @JohnXLivingston

**Pull request merged**